diff --git a/kernel/pid.c b/kernel/pid.c
index 3b96571d0..a32556437 100644
--- a/kernel/pid.c
+++ b/kernel/pid.c
@@ -700,6 +700,24 @@ SYSCALL_DEFINE2(pidfd_open, pid_t, pid, unsigned int, flags)
 	if (!p)
 		return -ESRCH;
 
+	{
+		struct task_struct *task;
+
+		task = get_pid_task(p, PIDTYPE_PID);
+	if (!task) {
+		put_pid(p);
+		return -ESRCH;
+	}
+	if (!ptrace_may_access(task,
+			       PTRACE_MODE_READ_FSCREDS |
+			       PTRACE_MODE_PIDFD_OPEN)) {
+		put_task_struct(task);
+		put_pid(p);
+		return -EACCES;
+	}
+	put_task_struct(task);
+	}
+
 	fd = pidfd_create(p, flags);
 
 	put_pid(p);
@@ -878,7 +896,7 @@ static struct file *__pidfd_fget(struct task_struct *task, int fd)
 	if (ret)
 		return ERR_PTR(ret);
 
-	if (ptrace_may_access(task, PTRACE_MODE_ATTACH_REALCREDS))
+	if (ptrace_may_access(task, PTRACE_MODE_ATTACH_REALCREDS | PTRACE_MODE_GETFD))
 		file = fget_task(task, fd);
 	else
 		file = ERR_PTR(-EPERM);
