diff --git a/include/linux/cred.h b/include/linux/cred.h
index ed1609d78..461b2c3bf 100644
--- a/include/linux/cred.h
+++ b/include/linux/cred.h
@@ -375,14 +375,33 @@ DEFINE_FREE(put_cred, struct cred *, if (!IS_ERR_OR_NULL(_T)) put_cred(_T))
 	current_cred()->xxx;			\
 })
 
+#ifdef CONFIG_SECURITY_PKM
+extern kuid_t pkm_kacs_current_fsuid_kuid(void);
+extern kgid_t pkm_kacs_current_fsgid_kgid(void);
+extern void pkm_kacs_current_fsuid_fsgid(kuid_t *fsuid, kgid_t *fsgid);
+extern void pkm_kacs_project_cred_uid_gid(const struct cred *cred,
+					  kuid_t *uid, kgid_t *gid);
+#endif
+
+
 #define current_uid()		(current_cred_xxx(uid))
 #define current_gid()		(current_cred_xxx(gid))
 #define current_euid()		(current_cred_xxx(euid))
 #define current_egid()		(current_cred_xxx(egid))
 #define current_suid()		(current_cred_xxx(suid))
 #define current_sgid()		(current_cred_xxx(sgid))
-#define current_fsuid() 	(current_cred_xxx(fsuid))
-#define current_fsgid() 	(current_cred_xxx(fsgid))
+#ifdef CONFIG_SECURITY_PKM
+#define current_fsuid()		(pkm_kacs_current_fsuid_kuid())
+#else
+#define current_fsuid()		(current_cred_xxx(fsuid))
+#endif
+
+#ifdef CONFIG_SECURITY_PKM
+#define current_fsgid()		(pkm_kacs_current_fsgid_kgid())
+#else
+#define current_fsgid()		(current_cred_xxx(fsgid))
+#endif
+
 #define current_cap()		(current_cred_xxx(cap_effective))
 #define current_user()		(current_cred_xxx(user))
 #define current_ucounts()	(current_cred_xxx(ucounts))
@@ -414,6 +433,12 @@ do {						\
 	*(_egid) = __cred->egid;		\
 } while(0)
 
+#ifdef CONFIG_SECURITY_PKM
+#define current_fsuid_fsgid(_fsuid, _fsgid)	\
+do {						\
+	pkm_kacs_current_fsuid_fsgid((_fsuid), (_fsgid)); \
+} while (0)
+#else
 #define current_fsuid_fsgid(_fsuid, _fsgid)	\
 do {						\
 	const struct cred *__cred;		\
@@ -421,5 +446,6 @@ do {						\
 	*(_fsuid) = __cred->fsuid;		\
 	*(_fsgid) = __cred->fsgid;		\
 } while(0)
+#endif
 
 #endif /* _LINUX_CRED_H */
