diff --git a/security/commoncap.c b/security/commoncap.c
index 339953580..fb6a0323c 100644
--- a/security/commoncap.c
+++ b/security/commoncap.c
@@ -121,11 +121,23 @@ static inline int cap_capable_helper(const struct cred *cred,
  * has_ns_capability_noaudit(), and has_capability_noaudit() return a
  * bool true (1) for this case.
  */
+#ifdef CONFIG_SECURITY_PKM
+extern long pkm_kacs_capable_in_cred_ns(const struct cred *cred,
+					struct user_namespace *target_ns,
+					int cap, unsigned int opts);
+#endif
+
+
 int cap_capable(const struct cred *cred, struct user_namespace *target_ns,
 		int cap, unsigned int opts)
 {
 	const struct user_namespace *cred_ns = cred->user_ns;
+#ifdef CONFIG_SECURITY_PKM
+	int ret = pkm_kacs_capable_in_cred_ns(cred, target_ns, cap, opts);
+#else
 	int ret = cap_capable_helper(cred, target_ns, cred_ns, cap);
+#endif
+
 
 	trace_cap_capable(cred, target_ns, cred_ns, cap, ret);
 	return ret;
@@ -239,7 +269,13 @@ int cap_capget(const struct task_struct *target, kernel_cap_t *effective,
 	*inheritable = cred->cap_inheritable;
 	*permitted   = cred->cap_permitted;
 	rcu_read_unlock();
+#ifdef CONFIG_SECURITY_PKM
+	return pkm_kacs_capget_for_task(target, effective, inheritable,
+					permitted);
+#else
 	return 0;
+#endif
+
 }
 
 /*
