diff --git a/arch/x86/entry/syscalls/syscall_64.tbl b/arch/x86/entry/syscalls/syscall_64.tbl
index 524155d65..f7c001119 100644
--- a/arch/x86/entry/syscalls/syscall_64.tbl
+++ b/arch/x86/entry/syscalls/syscall_64.tbl
@@ -441,3 +441,28 @@
 547	x32	pwritev2		compat_sys_pwritev64v2
 # This is the end of the legacy x32 range.  Numbers 548 and above are
 # not special and are not to be used for x32-specific syscalls.
+1000	common	kacs_open_self_token		sys_kacs_open_self_token
+1001	common	kacs_open_process_token		sys_kacs_open_process_token
+1002	common	kacs_open_thread_token		sys_kacs_open_thread_token
+1003	common	kacs_create_token		sys_kacs_create_token
+1004	common	kacs_create_session		sys_kacs_create_session
+1005	common	kacs_set_psb		sys_kacs_set_psb
+1006	common	kacs_destroy_empty_session		sys_kacs_destroy_empty_session
+1010	common	kacs_open_peer_token		sys_kacs_open_peer_token
+1011	common	kacs_impersonate_peer		sys_kacs_impersonate_peer
+1012	common	kacs_revert		sys_kacs_revert
+1013	common	kacs_set_impersonation_level		sys_kacs_set_impersonation_level
+1020	common	kacs_open		sys_kacs_open
+1021	common	kacs_get_sd		sys_kacs_get_sd
+1022	common	kacs_set_sd		sys_kacs_set_sd
+1023	common	kacs_access_check		sys_kacs_access_check
+1024	common	kacs_access_check_list		sys_kacs_access_check_list
+1025	common	kacs_set_caap		sys_kacs_set_caap
+1026	common	kacs_get_mount_policy		sys_kacs_get_mount_policy
+1027	common	kacs_set_mount_policy		sys_kacs_set_mount_policy
+1090	common	kmes_emit		sys_kmes_emit
+1091	common	kmes_attach		sys_kmes_attach
+1092	common	kmes_emit_batch		sys_kmes_emit_batch
+1100	common	reg_open_key		sys_reg_open_key
+1101	common	reg_create_key		sys_reg_create_key
+1102	common	reg_begin_transaction		sys_reg_begin_transaction
