diff --git a/fs/open.c b/fs/open.c
index 91f1139..73e5097 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -386,6 +386,7 @@ COMPAT_SYSCALL_DEFINE6(fallocate, int, fd, int, mode, compat_arg_u64_dual(offset
  * Creating new credentials is expensive, so we try to skip doing it,
  * which we can if the result would match what we already got.
  */
+#ifndef CONFIG_SECURITY_PKM
 static bool access_need_override_creds(int flags)
 {
 	const struct cred *cred;
@@ -461,6 +462,7 @@ static const struct cred *access_override_creds(void)
 	override_cred->non_rcu = 1;
 	return override_creds(override_cred);
 }
+#endif
 
 static int do_faccessat(int dfd, const char __user *filename, int mode, int flags)
 {
@@ -479,11 +481,13 @@ static int do_faccessat(int dfd, const char __user *filename, int mode, int flag
 	if (flags & AT_SYMLINK_NOFOLLOW)
 		lookup_flags &= ~LOOKUP_FOLLOW;
 
+#ifndef CONFIG_SECURITY_PKM
 	if (access_need_override_creds(flags)) {
 		old_cred = access_override_creds();
 		if (!old_cred)
 			return -ENOMEM;
 	}
+#endif
 
 	CLASS(filename_uflags, name)(filename, flags);
 retry:
@@ -503,6 +507,11 @@ retry:
 			goto out_path_release;
 	}
 
+#ifdef CONFIG_SECURITY_PKM
+	res = pkm_kacs_path_access(&path, mode);
+	if (res)
+		goto out_path_release;
+#endif
 	res = inode_permission(mnt_idmap(path.mnt), inode, mode | MAY_ACCESS);
 	/* SuS v2 requires we report a read only fs too */
 	if (res || !(mode & S_IWOTH) || special_file(inode->i_mode))
